File: common/history/src/createHistory.ts
import produce, { castImmutable } from "immer";
import { History } from "./History";

/**
 *
 * @param state state must be immutable because createHistory will freeze it.
 */
export function createHistory<TState>(state: TState): History<TState> {
  const freezedState = castImmutable(produce(state, () => {}));
  return {
    currentIndex: 0,
    states: [freezedState],
  };
}

File: rust-test/history/src/create_history.rs
use crate::History::History;

impl<TState> History<TState>
where
    TState: Copy,
{
    pub fn create_history(state: TState) -> History<TState> {
        History {
            current_index: 0,
            states: vec![state],
        }
    }
}

#[cfg(test)]
mod tests {}
